src/tools/dashboard/src/reference.rs

```diff
@@ -0,0 +1,287 @@
+                // for the end of the link tag before adding the path and
+                // hierarchy of the current chapter/section to the map.
+                map.insert(path.split('/').collect(), hierarchy.clone());
+            }
+            pulldown_cmark::Event::Text(text) => {
+                // Add the current chapter/section title to the hierarchy.
+                hierarchy.push(text.to_string());
+            }
+            _ => (),
+        }
+    }
+    map
+}
+
+/// Extracts examples from the given relative `paths` in the `book_dir` and
+/// saves them in `gen_dir`.
+fn extract_examples(paths: Vec<&PathBuf>, book_dir: &Path, gen_dir: &Path) {
+    for path in paths {
+        let mut cmd = Command::new("rustdoc");
+        cmd.args([
+            "+nightly",
+            "--test",
+            "-Z",
+            "unstable-options",
+            book_dir.join(path).to_str().unwrap(),
+            "--test-builder",
+            &["src", "tools", "dashboard", "print.sh"]
+                .iter()
+                .collect::<PathBuf>()
+                .to_str()
+                .unwrap(),
+            "--persist-doctests",
+            gen_dir.to_str().unwrap(),
+            "--no-run",
+        ]);
+        cmd.stdout(Stdio::null());
+        cmd.spawn()
+            .expect(&format!("Error: failed to spawn process:\n{:?}", cmd))
+            .wait()
+            .expect(&format!("Error: spawned process did not run:\n{:?}", cmd));
+    }
+}
+
+/// Copies the extracted rust code in `from_dir` to `src/test` following the
+/// hierarchy specified by `rel_map`.
+fn organize_examples(rel_map: &HashMap<PathBuf, PathBuf>, book_dir: &Path, from_dir: &Path) {
+    let to_dir: PathBuf = ["src", "test"].iter().collect();
+    // The names of the extracted examples generated by `rustdoc` have the
+    // format `<path>_<line-num>_<test-num>` where occurrences of '/', '-', and
+    // '.' in <path> are replaced by '_'. This translation is not injective, so
+    // we cannot map those names to the original markdown file path. Instead, we
+    // create a new map that does the same translation.
+    let mut abs_map = HashMap::new();
+    for (path, hierarchy) in rel_map.iter() {
+        abs_map.insert(
+            book_dir.join(path).to_str().unwrap().replace(&['\\', '/', '-', '.'][..], "_"),
+            to_dir.join(hierarchy),
+        );
+    }
+    for dir in from_dir.read_dir().unwrap() {
+        let dir = dir.unwrap().path();
+        // Some directories do not contain tests because the markdown file
+        // instructs `rustdoc` to ignore those tests.
+        if let Some(example) = dir.read_dir().unwrap().next() {
+            let example = example.unwrap().path();
+            copy(&example, &abs_map);
+        }
+    }
+}
+
+/// Copy the file specified by `from` to the corresponding location specified by
+/// `map`.
+fn copy(from: &Path, map: &HashMap<String, PathBuf>) {
```

Comment:
Could you include a comment with the full pattern of what `from` looks like? I'm having trouble piecing together these `find`s.

Reply:
Done! I rewrote the whole function too. Take another look and let me know if it makes sense to you now.